### ISSUE TITLE
[feature request] About Nan error using acos() function

### ISSUE THREAD

**Issue by syinari0123** [pytorch/pytorch#8069]
I created a network using torch.acos() and encountered a Nan error during its training.

The reason is that an input value to torch.acos () approaches 1 or -1, causing the grad to diverge, and resulting in value.grad = Nan.

I know that torch.acos () can only handle values ​​in the range from -1 to 1, and limited the input value's range from -1 to 1 using clamp(min=-1,max=1). 
However this error has occurred.
Finally I solved this by limiting input from -1 + eps to 1-eps with eps = 1e-7.

I think that many users would encounter this error.
I suggest "to alert this in the explanation of torch.acos() function in document" or "to change function to explicitly limit the range of values ​​by including eps as argument".

**Comment by Solacex** [comment]
Thanks a lot !

**Comment by laycoding** [comment]
Thanks a lot !

**Comment by KeremTurgutlu** [comment]
Thank you !

**Comment by maxjiang93** [comment]
Thanks! Would have been nice if they adopted your feature request. Desperately finding NaNs in networks half-way through the training is not fun... I learned this the hard way... The bug boils down to acos() function of 1.

**Comment by hackerpeter1** [comment]
666

**Comment by colesbury** [comment]
You should use https://pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly to help find where NaNs occur in your computation.

We're not going to change the behavior of torch.acos but PRs improving the documentation are very welcome.

**Comment by HapeMask** [comment]
Since this is currently the top google result for "pytorch acos nan grad" I figured I'd share a temporary workaround / hack to unblock people:

```python
def acos_safe(x, eps=1e-4):
    slope = np.arccos(1-eps) / eps
    # TODO: stop doing this allocation once sparse gradients with NaNs (like in
    # th.where) are handled differently.
    buf = th.empty_like(x)
    good = abs(x) <= 1-eps
    bad = ~good
    sign = th.sign(x[bad])
    buf[good] = th.acos(x[good])
    buf[bad] = th.acos(sign * (1 - eps)) - slope*sign*(abs(x[bad]) - 1 + eps)
    return buf
```

If #16329 is resolved, it would become much shorter:

```python
def acos_safe(x, eps=1e-4):
    sign = th.sign(x)
    slope = np.arccos(1-eps) / eps
    return th.where(abs(x) <= 1-eps,
                    th.acos(x),
                    th.acos(sign * (1 - eps)) - slope*sign*(abs(x) - 1 + eps))
```

**Comment by RSKothari** [comment]
@HapeMask I ran a test case on your code snippet.

```
import torch as th
import numpy as np
A = th.ones(5,5) + 0.1
acos_safe(A)
```
The output:
![image](https://user-images.githubusercontent.com/10388010/87679562-5709f880-c74a-11ea-82fb-c5a7c89dd48a.png)

**Comment by HapeMask** [comment]
> @HapeMask I ran a test case on your code snippet. I'm not sure if this is the behaviour you're expecting from `acos`.
> 
> ```
> import torch as th
> import numpy as np
> A = th.ones(5,5) + 0.1
> acos_safe(A)
> ```
> 
> The output:
> ![image](https://user-images.githubusercontent.com/10388010/87679562-5709f880-c74a-11ea-82fb-c5a7c89dd48a.png)

That’s expected, the function is only defined on [-1, 1] and this is simply a linear extrapolation outside the bounds.

**Comment by athn-nik** [comment]
Here is another quick workout which got me through this:
```
epsilon=1e-8 
torch.acos(torch.clamp(array, -1 + epsilon, 1 - epsilon))
```

**Comment by mpdprot** [comment]
@athn-nik thanks for the tip! I was using `float32` data types, so `1e-8` was too small. The minimum value of epsilon that worked for me was `3e-8`, anything lower than that and `acos` evaluates to `0.` at `1-epsilon`. `1e-7` worked out to be a good value.

**Comment by mpdprot** [comment]
@athn-nik thanks for the tip! I was using `float32` data types, so `1e-8` was too small. The minimum value of epsilon that worked for me was `3e-8`, anything lower than that and `torch.acos(1-epsilon)` evaluates to `0.`

`epsilon=1e-7` worked out to be a good value.

**Comment by jungerm2** [comment]
Any progress on this? The epsilon hack works well enough, I'm just surprised this issue hasn't gained more attention...

**Comment by lix4** [comment]
In CPU mode, I got `nan`. However, this issue disappear after I put data into GPU.

**Comment by tannlonggtaoo** [comment]
I met exactly the same problem today.
I tried the epsilon clipping method, and it works good in my pose detection network...

**Comment by maxwass** [comment]
For a robust `acos` with stable gradients, address singularities at  $z = \pm 1$ by setting the gradient outside $(1 - \epsilon, 1 + \epsilon)$ to $\left.\frac{d}{dz}arc\cos(z)\right|_{z = 1 - \epsilon}$. This can be particularly useful for e.g. differentiating distances on hyperspheres.

![image](https://github.com/pytorch/pytorch/assets/9613809/92c5cf49-0c57-4f97-99a2-153b9702c086)

```python
import torch

def acos_grad(z):
    return -1 / torch.sqrt(1 - z**2)

class safe_acos(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        ctx.save_for_backward(input)
        
        # protect ourselves from nan outputs in forward pass.
        return torch.clamp(input, min=-1 + 1e-6, max=1 - 1e-6).acos()

    @staticmethod
    def backward(ctx, grad_output):
        input, = ctx.saved_tensors
        grad_input = grad_output.clone()

        # protect ourselves from large gradients in backward pass.
        # outside of (-1 + epsilon, 1 - epsilon), gradient value is fixed constant to acos'(1-epsilon)
        epsilon = .05
        safe_input = torch.clamp(input, min=-1 + epsilon, max=1 - epsilon)

        return acos_grad(safe_input) * grad_input


np.set_printoptions(precision=4)
np.set_printoptions(suppress=True)

d = 1e-3
z = torch.tensor([-1.0 - d, -1.00 + d, 0.5, 1.0 - d, 1.00 + d], requires_grad=True)
print('inputs:          ', z.detach().numpy())
print('torch acos:      ', torch.acos(z).detach().numpy()) # raw acos returns nan outside of range
y = safe_acos.apply(z)
print('safe acos:       ', y.detach().numpy())

print('torch acos grad: ', acos_grad(z).detach().numpy()) # large gradient near the boundaries
y.backward(torch.ones_like(y))
print('safe acos grad:  ', z.grad.detach().numpy())

# inputs:           [-1.001 -0.999  0.5    0.999  1.001]
# torch acos:       [   nan 3.0969 1.0472 0.0447    nan]
# safe acos:        [3.1402 3.0969 1.0472 0.0447 0.0014]
# torch acos grad:  [     nan -22.3665  -1.1547 -22.3665      nan]
# safe acos grad:   [-3.2026 -3.2026 -1.1547 -3.2026 -3.2026]
```
